### ISSUE TITLE
Maven archetype project debug fails with "Timed out waiting for debug server to start"

### ISSUE THREAD

**Issue by agentHoover** [kirksl/karate-runner#70]
I create a new maven project using the karate archetype.  I followed [the instructions](https://github.com/kirksl/karate-runner#debugger-1) for creating a launch.json configuration for maven but when I try to start a debug session from the code lens link in the `users.feature` example test it eventually times out waiting for the debug server to start.  Is there something I am missing here to get this working?

```
{
    "configurations": [
        {
            "type": "karate",
            "name": "Karate (debug): Maven",
            "request": "launch",
            "feature": "${command:karateRunner.getDebugFile}",
            "karateOptions": "",
            "karateCli": "mvn test-compile -f \"${command:karateRunner.getDebugBuildFile}\" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args=\"-d\" -Dexec.classpathScope=\"test\" ${config:karateRunner.karateRunner.commandLineArgs}"
        }
    ]
}
```
The console output indicates that the debug server has started but the extension fails to connect 
```
Executing task: mvn test-compile -f "/home/moose/Data/Dev/ToolSmackDown/karate2/karatedemo/pom.xml" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args="-d" -Dexec.classpathScope="test"  <

[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------< org.improving:karatedemo >----------------------
[INFO] Building karatedemo 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ karatedemo ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/moose/Data/Dev/ToolSmackDown/karate2/karatedemo/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.0:compile (default-compile) @ karatedemo ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ karatedemo ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.0:testCompile (default-testCompile) @ karatedemo ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:java (default-cli) @ karatedemo ---
command: -d
10:51:38.519 [com.intuit.karate.cli.Main.main()] INFO  com.intuit.karate.debug.DapServer - debug server started on port: 36875
```
![vsWaitingForServer](https://user-images.githubusercontent.com/4604810/96612463-635e9e00-12c3-11eb-98a2-08053e286e9f.png)
![TimedOutWaiting](https://user-images.githubusercontent.com/4604810/96612473-66598e80-12c3-11eb-8cb9-a494625afcd8.png)

**Comment by kirksl** [comment]
@agentHoover can you try reproducing this problem with the example project here.  https://github.com/kirksl/karate-maven-gradle

**Comment by agentHoover** [comment]
@kirksl Well, the short is yes, sort of.  But with some more experimentation I was able to get it to work intermittently.  After quite a bit of tinkering I have been able to reproduce this consistently with the following scenario. 
1. Open VSCode
2. Open feature file
3. Debug feature file from codelens. (debug is successful)
4. Run feature file from codelens (note I have override karate runner enable.  Tests won't run for me unless I turn this on.  That's another issue). 
5. Debug feature file from codelens 
Debug now fails and will continue to fail until I restart VSCode.  

Something about just running a feature file (via code lens or with the following command from the terminal) is breaking the debugging.  I've attached my project if that helps. 

`mvn clean test-compile -f "/home/moose/Data/Dev/ToolSmackDown/karate2/karatedemo/pom.xml" exec:java -Dexec.mainClass="com.intuit.karate.cli.Main" -Dexec.args="/home/moose/Data/Dev/ToolSmackDown/karate2/karatedemo/src/test/java/examples/users/users.feature" -Dexec.classpathScope="test"
`

[karatedemo.zip](https://github.com/kirksl/karate-runner/files/5419035/karatedemo.zip)

**Comment by tongilcoto** [comment]
Hi, I arrive here looking for an answer for the opposite problem. I can not run "Debug" Codelens from "karate-demo" project inside "karate" repo.
In any case, I have run the example project for "karate-maven-gradle" repo and I got this error when running "Debug" Codelens (java jdk 11)

```
> Executing task: mvn test-compile -f "/Users/agilcoto/Work/scratches/karate-maven-gradle/pom.xml" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args="-d" -Dexec.classpathScope="test"  <

[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.company:karate-maven-gradle >-------------------
[INFO] Building karate-maven-gradle 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:prepare-agent (default-prepare-agent) @ karate-maven-gradle ---
[INFO] argLine set to -javaagent:/Users/agilcoto/.m2/repository/org/jacoco/org.jacoco.agent/0.8.4/org.jacoco.agent-0.8.4-runtime.jar=destfile=/Users/agilcoto/Work/scratches/karate-maven-gradle/target/jacoco.exec
[INFO] 
[INFO] --- maven-resources-plugin:2.7:resources (default-resources) @ karate-maven-gradle ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ karate-maven-gradle ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.7:testResources (default-testResources) @ karate-maven-gradle ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 4 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ karate-maven-gradle ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- exec-maven-plugin:1.5.0:java (default-cli) @ karate-maven-gradle ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.459 s
[INFO] Finished at: 2020-11-17T09:10:05+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java (default-cli) on project karate-maven-gradle: The parameters 'mainClass' for goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java are missing or invalid -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginParameterException
The terminal process "/bin/zsh '-c', 'mvn test-compile -f "/Users/agilcoto/Work/scratches/karate-maven-gradle/pom.xml" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args="-d" -Dexec.classpathScope="test" '" terminated with exit code: 1.

Terminal will be reused by tasks, press any key to close it.
```


When running "Run" CodeLens, it success

**Comment by tongilcoto** [comment]
My problem with "Debug" CodeLens it is not just it is not detecting the debug server, but not getting the localhost:8080 url

Following my investigations around "karate-demo" project at "karate" repo, this is the log following @agentHoover procedure

```
> Executing task: mvn test-compile -f "/Users/agilcoto/Work/scratches/karate/karate-demo/pom.xml" exec:java -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args="-d" -Dexec.classpathScope="test"  <

[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for com.intuit.karate:karate-demo:jar:0.9.6
[WARNING] 'dependencies.dependency.scope' for org.springframework.boot:spring-boot-dependencies:pom must be one of [provided, compile, runtime, test, system] but is 'import'. @ line 28, column 20
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO] 
[INFO] -------------------< com.intuit.karate:karate-demo >--------------------
[INFO] Building karate-demo 0.9.6
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ karate-demo ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 6 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.0:compile (default-compile) @ karate-demo ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ karate-demo ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 164 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.0:testCompile (default-testCompile) @ karate-demo ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:java (default-cli) @ karate-demo ---
command: -d
10:07:14.114 [com.intuit.karate.cli.Main.main()] INFO  com.intuit.karate.debug.DapServer - debug server started on port: 64165
10:07:15.471 [nioEventLoopGroup-3-1] DEBUG c.i.karate.debug.DapServerHandler - command line: /Users/agilcoto/Work/scratches/karate/karate-demo/src/test/java/driver/core/test02.feature:3
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by net.minidev.asm.DynamicClassLoader (file:/Users/agilcoto/.m2/repository/net/minidev/accessors-smart/1.2/accessors-smart-1.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)
WARNING: Please consider reporting this to the maintainers of net.minidev.asm.DynamicClassLoader
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Warning: Nashorn engine is planned to be removed from a future JDK release
10:07:16.614 [Thread-1] INFO  com.intuit.karate.Runner - waiting for parallel features to complete ...
10:07:16.626 [pool-1-thread-1] INFO  com.intuit.karate - found scenario at line: 3
10:07:17.018 [ForkJoinPool-1-worker-3] INFO  com.intuit.karate - >> lock acquired, begin callSingle: classpath:demo/headers/common-noheaders.feature
10:07:17.515 [ForkJoinPool-1-worker-3] DEBUG com.intuit.karate - request:
1 > GET http://127.0.0.1:8080/headers
1 > Accept-Encoding: gzip,deflate
1 > Connection: Keep-Alive
1 > Host: 127.0.0.1:8080
1 > User-Agent: Apache-HttpClient/4.5.12 (Java/11.0.9)

10:07:17.543 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused), http call failed after 28 milliseconds for URL: http://127.0.0.1:8080/headers
10:07:17.543 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - http request failed: 
org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused)
10:07:17.545 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - feature call failed: classpath:demo/headers/common-noheaders.feature
arg: {"demoBaseUrl":"http://127.0.0.1:8080"}
common-noheaders.feature:8 - 
org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused)
10:07:17.547 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - javascript function call failed: com.intuit.karate.exception.KarateException: 
common-noheaders.feature:8 - 
org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused)
10:07:17.547 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - failed function body: function() {
  karate.configure('connectTimeout', 5000);
  karate.configure('readTimeout', 5000);  
  var port = karate.properties['demo.server.port'];  
  if (!port) {
    port = karate.env == 'web' ? 8090 : 8080;
  }
  var protocol = 'http';
  if (karate.properties['demo.server.https'] == 'true') {
    protocol = 'https';
    karate.configure('ssl', true);
  }  
  var config = { demoBaseUrl: protocol + '://127.0.0.1:' + port };
  if (karate.env == 'proxy') {
    var proxyPort = karate.properties['demo.proxy.port'];
    karate.configure('proxy', 'http://127.0.0.1:' + proxyPort);
  }
  if (karate.env != 'mock' && karate.env != 'proxy' && karate.env != 'contract') {
    // karate.configure('callSingleCache', { minutes: 1 });
    // 'callSingle' is guaranteed to run only once even across all threads
    var result = karate.callSingle('classpath:demo/headers/common-noheaders.feature', config);
    // and it sets a variable called 'authInfo' used in headers-single.feature
    config.authInfo = { authTime: result.time, authToken: result.token };
  }
  return config;
}
10:07:17.548 [ForkJoinPool-1-worker-3] ERROR com.intuit.karate - evaluation of 'karate-config.js' failed: javascript function call failed: com.intuit.karate.exception.KarateException: 
common-noheaders.feature:8 - 
org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused)
10:07:17.664 [pool-1-thread-1] INFO  com.intuit.karate.Runner - <<fail>> feature 1 of 1: karate-demo/src/test/java/driver/core/test02.feature
---------------------------------------------------------
feature: karate-demo/src/test/java/driver/core/test02.feature:3
report: target/surefire-reports/karate-demo.src.test.java.driver.core.test02.json
scenarios:  1 | passed:  0 | failed:  1 | time: 0,0000
---------------------------------------------------------

HTML report: (paste into browser to view) | Karate version: 0.9.6
file:/Users/agilcoto/Work/scratches/karate/target/surefire-reports/karate-summary.html
===================================================================

Karate version: 0.9.6
======================================================
elapsed:   2,18 | threads:    1 | thread time: 0,00 
features:     1 | ignored:    0 | efficiency: 0,00
scenarios:    1 | passed:     0 | failed: 1
======================================================
failed features:
karate-demo.src.test.java.driver.core.test02: test02.feature:3 - evaluation of 'karate-config.js' failed: javascript function call failed: com.intuit.karate.exception.KarateException: 
common-noheaders.feature:8 - 
org.apache.http.conn.HttpHostConnectException: Connect to 127.0.0.1:8080 [/127.0.0.1] failed: Connection refused (Connection refused)

10:07:17.759 [Thread-1] INFO  com.intuit.karate.debug.DapServer - stop: shutting down
10:07:17.765 [Thread-1] INFO  com.intuit.karate.debug.DapServer - stop: shutdown complete

Terminal will be reused by tasks, press any key to close it.
```

Same as previous, when running "Run" CodeLens, it successes

**Comment by kirksl** [comment]
@ptrthomas appreciate it if you can confirm why this error might be occurring for @tongilcoto .  if i rename pom.xml in the project below to force gradle, debug server is working just fine.  suspecting either the maven command or pom.xml might be an issue?  i tried to add the plugin in the error and that did not resolve the issue.


calling command
mvn test-compile exec:java -e -Dexec.mainClass=com.intuit.karate.cli.Main -Dexec.args=-d -Dexec.classpathScope=test

suggested here
https://github.com/intuit/karate/wiki/Debug-Server#maven

against project
https://github.com/kirksl/karate-maven-gradle

we see this error
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java (default-cli) on project karate-maven-gradle: The parameters 'mainClass' for goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java are missing or invalid -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java (default-cli) on project karate-maven-gradle: The parameters 'mainClass' for goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java are missing or invalid

**Comment by ptrthomas** [comment]
@kirksl pretty sure that parent pom in your project is to blame, you never see this with other maven projects

this needs some digging. can you try to create a pom without the parent pom / bom - it is possible and needs you to explicitly list versions etc, this is an example: https://github.com/intuit/karate/blob/master/examples/consumer-driven-contracts/payment-producer/pom.xml

**Comment by kirksl** [comment]
@ptrthomas many thanks the parent spring boot pom was indeed causing a problem.  
@tongilcoto please try the 'karate-maven-gradle' project again.  i was able to update pom.xml to enable "karate: debug" to work under the following conditions
1. update your karate command line to run in env=stg
![image](https://user-images.githubusercontent.com/24948266/99930477-0319ac80-2d06-11eb-8e1e-646c1f470cef.png)

2. prior to clicking on "karate: debug", start spring in another console with command "mvn clean spring-boot:run"

not quite sure why "karate: debug" doesn't work when using the mock.  using the same mock in gradle works perfectly fine.  i will need someone with better maven experience than me to help solve this part.

hope this helps.

**Comment by kirksl** [comment]
@agentHoover please see my prior comment.  hoping this well help resolve your initial issue as well?

**Comment by kirksl** [comment]
closing due to no response.  please open new issue if this is still a problem.
